File: python/subscription_manager/services.py
from datetime import datetime, timedelta
from typing import Dict, Callable
from .models import Client, Subscription, Plan
from .storage import StorageBackend
from .exceptions import TrialAlreadyUsed, PlanNotFound

class SubscriptionService:
    def __init__(
        self,
        storage: StorageBackend,
        plans: Dict[str, Plan],
        id_generator: Callable[[], str],
    ):
        self.storage = storage
        self.plans = plans
        self.id_generator = id_generator

    def register_client(self, email: str, phone: str = None, fingerprint: str = None) -> Client:
        existing = self.storage.get_client_by_email(email)
        if existing:
            return existing
        client = Client(
            id=self.id_generator(),
            email=email,
            phone=phone,
            fingerprint=fingerprint
        )
        self.storage.add_client(client)
        return client

    def start_subscription(self, client: Client, plan_key: str) -> Subscription:
        if plan_key not in self.plans:
            raise PlanNotFound(f"No such plan: {plan_key}")

        plan = self.plans[plan_key]

        if plan_key == "free_trial":
            subs = self.storage.get_subscriptions(client.id)
            if any(s.plan_key == "free_trial" for s in subs):
                raise TrialAlreadyUsed("Client has already used the free trial")

        now = datetime.utcnow()
        sub = Subscription(
            client_id=client.id,
            plan_key=plan_key,
            start_date=now,
            end_date=now + timedelta(days=plan.duration_days),
            active=True
        )
        self.storage.add_subscription(sub)
        return sub

    def renew_subscription(self, client: Client, plan_key: str) -> Subscription:
        sub = self.start_subscription(client, plan_key)
        return sub

    def cancel_subscription(self, client: Client, plan_key: str) -> None:
        # simplistic: mark active=False on all matching
        subs = self.storage.get_subscriptions(client.id)
        for s in subs:
            if s.plan_key == plan_key and s.active:
                s.active = False
        # you'd add update logic in storage if needed

    def is_active(self, client: Client, plan_key: str) -> bool:
        subs = self.storage.get_subscriptions(client.id)
        now = datetime.utcnow()
        return any(
            s.plan_key == plan_key and s.active and s.start_date <= now < s.end_date
            for s in subs
        )

File: javascript/src/services.js
import { TrialAlreadyUsed, PlanNotFound } from './exceptions.js';
import { Subscription } from './models.js';

export class SubscriptionService {
  /**
   * @param {*} storage  — any backend implementing the 4 async methods
   * @param {{[key:string]:Plan}} plans
   * @param {() => string} idGenerator
   */
  constructor(storage, plans, idGenerator) {
    this.storage = storage;
    this.plans = plans;
    this.idGenerator = idGenerator;
  }

  async registerClient(email, phone = null, fingerprint = null) {
    let client = await this.storage.getClientByEmail(email);
    if (client) return client;

    client = { id: this.idGenerator(), email, phone, fingerprint, createdAt: new Date() };
    await this.storage.addClient(client);
    return client;
  }

  async startSubscription(client, planKey) {
    const plan = this.plans[planKey];
    if (!plan) throw new PlanNotFound(`No such plan: ${planKey}`);

    if (planKey === 'free_trial') {
      const subs = await this.storage.getSubscriptions(client.id);
      if (subs.some(s => s.planKey === 'free_trial')) {
        throw new TrialAlreadyUsed('Free trial already used');
      }
    }

    const now = new Date();
    const endDate = new Date(now.getTime() + plan.durationDays * 24*60*60*1000);
    const sub = new Subscription({
      clientId: client.id,
      planKey,
      startDate: now,
      endDate,
      active: true
    });
    await this.storage.addSubscription(sub);
    return sub;
  }

  async renewSubscription(client, planKey) {
    return this.startSubscription(client, planKey);
  }

  async cancelSubscription(client, planKey) {
    const subs = await this.storage.getSubscriptions(client.id);
    subs.filter(s => s.planKey === planKey && s.active)
        .forEach(s => { s.active = false; });
    // Note: persistent update would require extra storage method
  }

  async isActive(client, planKey) {
    const now = new Date();
    const subs = await this.storage.getSubscriptions(client.id);
    return subs.some(s =>
      s.planKey === planKey &&
      s.active &&
      s.startDate <= now && now < s.endDate
    );
  }
}
